File: _leetcode_/2349. Design a Number Container System.cpp
#define fastio ios_base::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr)
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <cstring>
#include <climits>
#include <iomanip>
#include <sstream>
#include <fstream>
#include <functional>
using namespace std;

class NumberContainers {
public:
    NumberContainers() {}

    void change(int index, int number) {
        if(indexToNumbers.find(index) != indexToNumbers.end()){
            // index not mapped rn.
            // find curr number on that index.
            // erase that mapping from number to indice mapping.
            int currMappedNumber = indexToNumbers[index];
            // remove specific index from the set of indices.
            numberToIndices[currMappedNumber].erase(index); 
            // if there is now an empty set after erasing, clean up the key itself. the entire map entry.
            if(numberToIndices[currMappedNumber].empty()){
                numberToIndices.erase(currMappedNumber);
            }
        }
        indexToNumbers[index] = number;
        numberToIndices[number].insert(index);
    }

    int find(int number) {
       if(numberToIndices.find(number) != numberToIndices.end()){
            return *numberToIndices[number].begin();
       }
        return -1;
    }

private:
    // Map from number to set of indices for min index return, kinda like heap logic
    unordered_map<int, set<int>> numberToIndices;
    unordered_map<int, int> indexToNumbers;
};

/**
 * Your NumberContainers object will be instantiated and called as such:
 * NumberContainers* obj = new NumberContainers();
 * obj->change(index,number);
 * int param_2 = obj->find(number);
 */

File: _leetcode_/javascript/2349. Design a Number Container System.js
class NumberContainers {
    constructor() {
        this.numbersToIndices = new Map(); // all indices num x is at, so we can return smallest.
        // number, minQueue (set of indices)
        this.indexToNumber = new Map(); // insert or replace
    }
}

/** 
 * @param {number} index 
 * @param {number} number
 * @return {void}
 */
NumberContainers.prototype.change = function (index, number) {
        this.indexToNumber.set(index,number);
        const hasNumber = this.numbersToIndices.has(number);
        if(!hasNumber){
            this.numbersToIndices.set(number, new MinPriorityQueue());
            // datastructures-js
            // https://datastructures-js.info/docs/priority-queue
        }
        this.numbersToIndices.get(number).enqueue(index);
};

/** 
 * @param {number} number
 * @return {number}
 */
NumberContainers.prototype.find = function (number) {
    const indices = this.numbersToIndices.get(number);
    if (!indices) return -1;
    while(!indices.isEmpty()){
        const index = indices.front();
        if(this.indexToNumber.get(index) === number) return index
        indices.dequeue();
    }
    return -1
};

/** 
 * Your NumberContainers object will be instantiated and called as such:
 * var obj = new NumberContainers()
 * obj.change(index,number)
 * var param_2 = obj.find(number)
 */